### PR TITLE
rgw:send x-amz-version-id header when upload files

### PR DESCRIPTION
**Pull request by cfanz**
To be compatible with aws s3,  an x-amz-version-id header should be returned.

For atomic upload, RGWPutObj::version_id will stores the version-id either generated by rgw randomly or read from user.  For multipart upload, RGWCompleteMultipart::version_id will stores the version-id either generated by rgw randomly or read from user.  

Function send_respones() will send 'x-amz-version-id' header when version_id is not empty.

### CODE REVIEW

**Comment by ZVampirEM77** [comment]
@cfanz I think you have committed some disrelated msg in this PR. Is it right?

**Comment by cfanz** [comment]
@ZVampirEM77 Thanks for your advice.  I have clean it up.

**Comment by cbodley** [comment]
thanks @cfanz, this looks good. could you please add a s3test case for this? for reference, https://github.com/ceph/s3-tests/pull/157 was the test for x-amz-version-id on GetObj

**Comment by cfanz** [comment]
@cbodley I made a [s3-test-pr](https://github.com/ceph/s3-tests/pull/199) for this. But I'm not sure whether we should provide  test-cases for BUCKET_SUSPENDED and BUCKET_VERSIONS-SUSPENDED buckets. For BUCKET_SUSPENDED bucket, which is the default state, version-id should be empty, and this is satisfied by rgw. However for BUCKET_VERSIONS-SUSPENDED bucket, version-id should be equal to string "null", but rgw still returns an empty version-id instead.

File: src/rgw/rgw_rados.h [hunk 1]
```diff
@@ -3934,6 +3934,10 @@ class RGWPutObjProcessor_Atomic : public RGWPutObjProcessor_Aio
   void set_version_id(const string& vid) {
     version_id = vid;
   }
+
+  string get_version_id() const {
+    return version_id;
+  }
```
Comment (cbodley) [inline src/rgw/rgw_rados.h]: can return as `const string&`

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -1384,10 +1384,14 @@ void RGWPutObj_ObjStore_S3::send_response()
       dump_errno(s);
       dump_etag(s, etag);
       dump_content_length(s, 0);
+      if (!version_id.empty())
+        dump_header(s, "x-amz-version-id", version_id.c_str());
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc]: there's a helper function to make these a bit easier:
```c++
      dump_header_if_nonempty(s, "x-amz-version-id", version_id);
```

File: src/rgw/rgw_common.h [hunk 3]
```diff
@@ -1537,6 +1537,10 @@ struct rgw_obj_key {
     instance = i;
   }
 
+  string get_instance() const {
+    return instance;
+  }
```
Comment (cbodley) [inline src/rgw/rgw_common.h]: can return as `const string&` - but `instance` is public, so this function might be overkill

File: src/rgw/rgw_rest_s3.cc (hunk 2 above)
Comment (cfanz) [reply src/rgw/rgw_rest_s3.cc]: @cbodley Thanks for your advice. I have amended it.

**Comment by cbodley** [comment]
> However for BUCKET_VERSIONS-SUSPENDED bucket, version-id should be equal to string "null", but rgw still returns an empty version-id instead.

i tried this on aws to see how their version-suspended buckets work on PUT and GET, and found that their GET reply does include `x-amz-version-id: null`, but their PUT reply doesn't. so i think it's definitely worth adding test cases here

**Comment by cfanz** [comment]
@cbodley Totally agree.  Now the [s3-test-pr](https://github.com/ceph/s3-tests/pull/199) will check version-id on default-bucket, versioning-enabled-bucket and versioning-suspended-bucket when upload files.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2017-11-25_00:12:48-rgw-wip-kefu-testing-2017-11-25-0120-distro-basic-mira/

@cbodley  is this ready to merge?

**Comment by cbodley** [comment]
thanks @tchaikov

**Pull request merged**